// @prismatic-io/spectral/dist/generators/componentManifest/createTriggers.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createTriggers = void 0;
const path_1 = __importDefault(require("path"));
const getInputs_1 = require("./getInputs");
const getImports_1 = require("./getImports");
const helpers_1 = require("./helpers");
const createTemplate_1 = require("../utils/createTemplate");
const createTypeInterface_1 = require("../utils/createTypeInterface");
const createImport_1 = require("../utils/createImport");
const createTriggers = (_a) => __awaiter(void 0, [_a], void 0, function* ({ component, dryRun, verbose, sourceDir, destinationDir, }) {
    var _b, _c;
    if (verbose) {
        console.info("Creating triggers...");
    }
    const triggersIndex = yield renderTriggersIndex({
        imports: Object.entries((_b = component.triggers) !== null && _b !== void 0 ? _b : {}).map(([triggerKey, trigger]) => {
            return {
                import: (0, createImport_1.createImport)(trigger.key || triggerKey),
            };
        }),
        dryRun,
        verbose,
        sourceDir,
        destinationDir,
    });
    const triggers = yield Promise.all(Object.entries((_c = component.triggers) !== null && _c !== void 0 ? _c : {}).map((_a) => __awaiter(void 0, [_a], void 0, function* ([triggerKey, trigger]) {
        const inputs = (0, getInputs_1.getInputs)({
            inputs: trigger.inputs,
        });
        const imports = (0, getImports_1.getImports)({ inputs });
        return yield renderTrigger({
            trigger: {
                typeInterface: (0, createTypeInterface_1.createTypeInterface)(trigger.key || triggerKey),
                import: (0, createImport_1.createImport)(trigger.key || triggerKey),
                key: trigger.key || triggerKey,
                label: trigger.display.description,
                description: trigger.display.description,
                inputs,
            },
            dryRun,
            imports,
            verbose,
            sourceDir,
            destinationDir,
        });
    })));
    if (verbose) {
        console.info("");
    }
    return Promise.resolve({
        triggersIndex,
        triggers,
    });
});
exports.createTriggers = createTriggers;
const renderTriggersIndex = (_a) => __awaiter(void 0, [_a], void 0, function* ({ imports, dryRun, verbose, sourceDir, destinationDir, }) {
    return yield (0, createTemplate_1.createTemplate)({
        source: path_1.default.join(sourceDir, "triggers", "index.ts.ejs"),
        destination: path_1.default.join(destinationDir, "triggers", "index.ts"),
        data: {
            imports,
        },
        dryRun,
        verbose,
    });
});
const renderTrigger = (_a) => __awaiter(void 0, [_a], void 0, function* ({ dryRun, imports, trigger, verbose, sourceDir, destinationDir, }) {
    return yield (0, createTemplate_1.createTemplate)({
        source: path_1.default.join(sourceDir, "triggers", "trigger.ts.ejs"),
        destination: path_1.default.join(destinationDir, "triggers", `${trigger.import}.ts`),
        data: {
            helpers: helpers_1.helpers,
            imports,
            trigger,
        },
        dryRun,
        verbose,
    });
});
